lib/Dialect/TTIR/Transforms/HoistCPUOps.cpp

```diff
@@ -0,0 +1,296 @@
+             llvm::StringRef argValue, std::vector<std::string> &value) {
+    // Split the input string on commas
+    llvm::SmallVector<llvm::StringRef, 4> names;
+    argValue.split(names, ',');
+
+    value.assign(names.begin(), names.end());
+
+    return false; // Success
+  }
+
+  static std::string toString(const std::vector<std::string> &locList) {
+    std::string result;
+    for (const auto &loc : locList) {
+      result += loc + ",";
+    }
+    return result;
+  }
+
+  static void print(llvm::raw_ostream &os,
+                    const std::vector<std::string> &locList) {
+    os << "hoist-locs=" << toString(locList) << "\n";
+  }
+};
+
+class TTIRHoistTransform
+    : public impl::TTIRHoistTransformBase<TTIRHoistTransform> {
+public:
+  using impl::TTIRHoistTransformBase<
+      TTIRHoistTransform>::TTIRHoistTransformBase;
+
+  // MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(TTIRHoistTransform)
+
+  void runOnOperation() final {
+    mlir::ModuleOp moduleOp = getOperation();
+    assert(moduleOp != nullptr && "TTIRHoistTransform should run on ModuleOps, "
+                                  "somehow got something else!");
+
+    IRRewriter rewriter(&getContext());
+
+    auto loc = moduleOp->getLoc();
+
+    // TODO (vwells): add logic for automatic vs manual flag here when we have
+    // automatic analysis pass
+
+    // only create cpu-module etc. if we actually have ops to hoist
+    if (!hoistLocs.hasValue()) {
+      return;
+    }
+    const auto targetLocs = getLocations();
+    if (targetLocs.empty()) {
+      return;
+    }
+
+    TTIRHoistAnalyzeManual analysisPass(moduleOp, targetLocs);
+    const auto &hoistOpSets = analysisPass.getResults();
+
+    // Check if a "cpu_module" already exists
+    mlir::ModuleOp cpuModule;
+    for (auto &op : moduleOp.getBody()->getOperations()) {
+      if (auto module = llvm::dyn_cast<mlir::ModuleOp>(op)) {
+        if (module->hasAttr("ttir.cpu_module")) {
+          cpuModule = module;
+          break;
+        }
+      }
+    }
+
+    // If no CPU module exists, create one
+    if (!cpuModule) {
+      rewriter.setInsertionPointToEnd(moduleOp.getBody());
+      cpuModule = rewriter.create<mlir::ModuleOp>(loc);
+      cpuModule->setAttr("ttir.cpu_module", rewriter.getUnitAttr());
+      cpuModule->setAttr(mlir::SymbolTable::getSymbolAttrName(),
+                         rewriter.getStringAttr("cpu_module"));
+      // try to make cpu module global
+      mlir::SymbolTable::setSymbolVisibility(
+          cpuModule, mlir::SymbolTable::Visibility::Public);
+    }
+
+    for (const auto &opSet : hoistOpSets) {
+      assert(opSet.size() == 1 &&
+             "currently don't support hoisting multiple instructions at once!");
+      hoistOperationToFunction(*opSet.begin(), moduleOp, cpuModule);
+    }
+  }
+
+protected:
+  std::vector<std::string> getLocations() {
+    std::vector<std::string> locs;
+    for (const std::string &loc : hoistLocs) {
```

Comment:
Maybe return map here, so you can avoid looping, and has easier api for checking if op needs to be hoisted.

Reply:
Good idea, I will switch to LLVM SmallSet I think